docs/howto/data_rw.rst

```diff
@@ -0,0 +1,222 @@
+
+``pysdmx`` allows to return the written data as a string or write it to a file. SDMX-CSV writers only allow one dataset to be written at a time, while SDMX-ML writers allow multiple datasets to be written at once.
+
+SDMX-CSV 1.0
+^^^^^^^^^^^^
+
+`SDMX-CSV 1.0 specification <https://github.com/sdmx-twg/sdmx-csv/blob/v1.0/data-message/docs/sdmx-csv-field-guide.md>`_
+
+.. warning::
+
+        The SDMX-CSV 1.0 format is deprecated and should not be used for new implementations.
+        It only allows a dataflow to be represented, which is not enough for most use cases.
+
+.. autofunction:: pysdmx.io.csv.sdmx10.writer.writer
+
+.. code-block:: python
+
+   from pysdmx.io.csv.sdmx10.writer import writer
+   from pathlib import Path
+
+   # Write to file sample.csv in the same folder as this code
+   file_path = Path(__file__).parent / "sample.csv"
+   writer(dataset, file_path)
+
+
+SDMX-CSV 2.0
+^^^^^^^^^^^^
+
+`SDMX-CSV 2.0 specification <https://github.com/sdmx-twg/sdmx-csv/blob/v2.0.0/data-message/docs/sdmx-csv-field-guide.md>`_
+
+.. note::
+
+        The SDMX-CSV 2.0 writer will write the data as the `ordinary case <https://github.com/sdmx-twg/sdmx-csv/blob/v2.0.0/data-message/docs/sdmx-csv-field-guide.md#1-ordinary-case>`_. If you need to write data in other cases, you may need to write a custom script.
+
+.. warning::
+
+        We use only comma as the delimiter.
+
+.. autofunction:: pysdmx.io.csv.sdmx20.writer.writer
+
+.. code-block:: python
+
+   from pysdmx.io.csv.sdmx20.writer import writer
+   from pathlib import Path
+
+   # Write to file sample.csv in the same folder as this code
+   file_path = Path(__file__).parent / "sample.csv"
+   writer(dataset, file_path)
+
+SDMX-ML 2.1
+^^^^^^^^^^^
+
+SDMX-ML 2.1 format is described
+`here (pdf file for IM) <https://sdmx.org/wp-content/uploads/SDMX_2-1_SECTION_2_InformationModel_2020-07.pdf>`_
+
+SDMX-ML 2.1 format allows to write multiple datasets at once. To use the Series format, you need to pass the dimension at observation dictionary, where the key is the dataset short urn and the value is the dimension id to be observed.
+
+.. important::
+
+    For each dataset, if dataset.structure is not a Schema, the writer can only write in the Structure Specific All Dimensions format.
+    We perform a check to ensure that the dataset has a Schema structure for the remaining formats as we need to know the roles for each component. This check also ensures that the dataset.structure has at least one dimension and one measure defined.
+
+.. autofunction:: pysdmx.io.xml.sdmx21.writer.writer
+
+.. code-block:: python
+
+    from pysdmx.io.xml.sdmx21.writer import writer
+    from pysdmx.io.xml.enums import MessageType
+    from pathlib import Path
+
+    # Dictionary of datasets to write
+    datasets = {
+        "DataStructure=TEST_AGENCY:TEST_ID(1.0)": dataset
+    }
+
+    # Dimension at observation mapping
+    dim_mapping = {
+        "DataStructure=TEST_AGENCY:TEST_ID(1.0)": "TIME_PERIOD"
```

Comment:
The vast majority of available data out there uses TIME_PERIOD as period dimension. Therefore, mapping definition should not be necessary (and I think it isn't?). If so, maybe we could have the first example without dim_mapping and then add another explaining the use case behind it?

Reply:
Unfortunately at this point dim_mapping is necessary to generate the Series formats. IMHO I prefer to make this explicit so the user always selects the dimension for each dataset. We can as well include a change in the signature of the writer with a time_series flag to check if the dataset has the TIME_PERIOD dimension to write as Series, if not write as All_Dimensions